tests/libs/util/socket_helper.cpp

```diff
@@ -0,0 +1,326 @@
+    // Initialize the remote address.
+    ADDRINFO* address_info = nullptr;
+    // Try converting address string to IP address.
+    error = getaddrinfo(address_string.data(), nullptr, nullptr, &address_info);
+    if (error != 0)
+        FAIL("getaddrinfo for" << address_string << " failed with " << WSAGetLastError());
+    if (address_info->ai_family == AF_INET) {
+        IN6ADDR_SETV4MAPPED(
+            (PSOCKADDR_IN6)&address, (IN_ADDR*)INETADDR_ADDRESS(address_info->ai_addr), scopeid_unspecified, 0);
+    } else {
+        REQUIRE(address_info->ai_family == AF_INET6);
+        address.ss_family = AF_INET6;
+        INETADDR_SET_ADDRESS((PSOCKADDR)&address, INETADDR_ADDRESS(address_info->ai_addr));
+    }
+    if (address_family != nullptr)
+        *address_family = static_cast<ADDRESS_FAMILY>(address_info->ai_family);
+    freeaddrinfo(address_info);
+}
+
+_base_socket::_base_socket(int _sock_type, int _protocol, uint16_t _port)
+    : socket(INVALID_SOCKET), sock_type(_sock_type), protocol(_protocol), port(_port), local_address{},
+      local_address_size(sizeof(local_address))
+{
+    int error = 0;
+
+    socket = WSASocket(AF_INET6, sock_type, protocol, nullptr, 0, WSA_FLAG_OVERLAPPED);
+    if (socket == INVALID_SOCKET)
+        FAIL("Failed to create socket with error: " << WSAGetLastError());
+    uint32_t ipv6_opt = 0;
+    error = setsockopt(socket, IPPROTO_IPV6, IPV6_V6ONLY, reinterpret_cast<const char*>(&ipv6_opt), sizeof(ULONG));
+    if (error != 0)
+        FAIL("Could not enable dual family endpoint: " << WSAGetLastError());
+
+    // Bind it to the any address and supplied port.
+    SOCKADDR_STORAGE local_addr;
+    local_addr.ss_family = AF_INET6;
+    INETADDR_SETANY((PSOCKADDR)&local_addr);
+    ((PSOCKADDR_IN6)&local_addr)->sin6_port = htons(port);
+
+    error = bind(socket, (PSOCKADDR)&local_addr, sizeof(local_addr));
+    if (error != 0)
+        FAIL("Failed to bind socket with error: " << WSAGetLastError());
+
+    error = getsockname(socket, (PSOCKADDR)&local_address, &local_address_size);
+    if (error != 0)
+        FAIL("Failed to query local address of socket with error: " << WSAGetLastError());
+}
+
+_base_socket::~_base_socket() { closesocket(socket); }
+
+void
+_base_socket::get_local_address(PSOCKADDR& address, int& address_length)
+{
+    address = (PSOCKADDR)&local_address;
+    address_length = local_address_size;
+}
+
+_sender_socket::_sender_socket(int _sock_type, int _protocol, uint16_t _port)
+    : _base_socket{_sock_type, _protocol, _port}
+{}
+
+_datagram_sender_socket::_datagram_sender_socket(int _sock_type, int _protocol, uint16_t _port)
+    : _sender_socket{_sock_type, _protocol, _port}
+{
+    if (!(sock_type == SOCK_DGRAM || sock_type == SOCK_RAW) &&
+        !(protocol == IPPROTO_UDP || protocol == IPPROTO_IPV4 || protocol == IPPROTO_IPV6))
+        FAIL("datagram_socket only supports these combinations (SOCK_DGRAM, IPPROTO_UDP) and (SOCK_RAW, "
+             "IPPROTO_IPV4/IPV6)");
```

Comment:
this doesn't match what lines 80-81 do.
For example 80-81 will accept (SOCK_RAW, IPPROTO_UDP).

Reply:
Changed the log statement.